### PR TITLE
fix: optiming threads deadlock

### PR DESCRIPTION
**Pull request by cutecutecat**
- `optimizing_indexing` not ended due to no-ending monitor
- Fix a memory leak at `rayon.STOP`
- `crate-ci/typos` upgraded to v1.20 with many fake positive words fix: https://github.com/crate-ci/typos/issues/970

### CODE REVIEW

File: crates/rayon/src/lib.rs [hunk 1]
```diff
@@ -49,8 +49,15 @@ impl ThreadPoolBuilder {
         f: impl FnOnce(&ThreadPool),
     ) -> Result<(), rayon::ThreadPoolBuildError> {
         let stop = Arc::new(AtomicBool::new(false));
+        let stop_value = stop.clone();
```
[the remark this answers is not in the document]
Comment (cutecutecat) [reply crates/rayon/src/lib.rs:52]: Bad news, clippy doesn't like it: https://github.com/tensorchord/pgvecto.rs/actions/runs/8533026186/job/23375114326?pr=453

File: crates/index/src/optimizing/indexing.rs [hunk 2]
```diff
@@ -104,17 +103,20 @@ impl<O: Op> OptimizerIndexing<O> {
...
+                        let handler = scope.spawn(|| {
+                            monitor(&finish, &shutdown);
+                            pool.stop();
                         });
-                        let _ = pool.install(|| optimizing_indexing(index.clone()));
+                        let _ = pool.install(|| {
+                            let _ = optimizing_indexing(index.clone());
+                            drop(finish_tx);
+                        });
+                        let _ = handler.join();
```
[the remark this answers is not in the document]
Comment (cutecutecat) [reply crates/index/src/optimizing/indexing.rs:120]: Not worked

File: crates/rayon/src/lib.rs [hunk 3]
```diff
@@ -49,8 +49,15 @@ impl ThreadPoolBuilder {
...
         match std::panic::catch_unwind(AssertUnwindSafe(|| {
             self.builder
+                .start_handler(move |_| {
+                    STOP.replace(Some(stop_value.clone()));
+                })
```
Comment (usamoi) [inline crates/rayon/src/lib.rs:57]:
```suggestion
                .start_handler({
                    let stop = stop.clone();
                    move |_| {
                        STOP.replace(Some(stop.clone()));
                    }
                })
```

File: crates/rayon/src/lib.rs [hunk 4]
```diff
@@ -49,8 +49,18 @@ impl ThreadPoolBuilder {
         f: impl FnOnce(&ThreadPool),
     ) -> Result<(), rayon::ThreadPoolBuildError> {
         let stop = Arc::new(AtomicBool::new(false));
+        let stop_value = stop.clone();
```
Comment (usamoi) [inline crates/rayon/src/lib.rs:52]:
```suggestion
```

File: crates/index/src/optimizing/indexing.rs [hunk 5]
```diff
@@ -99,26 +99,29 @@ impl<O: Op> OptimizerIndexing<O> {
...
                 .build_scoped(|pool| {
                     std::thread::scope(|scope| {
-                        scope.spawn(|| match shutdown.recv() {
-                            Ok(never) => match never {},
-                            Err(RecvError) => {
-                                pool.stop();
-                            }
+                        let handler = scope.spawn(|| {
+                            monitor(&finish_rx, &shutdown_rx);
+                            pool.stop();
```
Comment (usamoi) [inline crates/index/src/optimizing/indexing.rs:114]: It should be called only if the reason why `monitor` returned is `shutdown_rx` disconnected.

**Review by usamoi — APPROVED** [review]

**Pull request merged**